src/aws_encryption_sdk/keyrings/aws_kms/__init__.py

```diff
@@ -0,0 +1,376 @@
+                    client_supplier=self._client_supplier,
+                    key_name=self._key_id,
+                    plaintext_data_key=encryption_materials.data_encryption_key,
+                    encryption_context=encryption_materials.encryption_context,
+                    grant_tokens=self._grant_tokens,
+                )
+        except Exception:  # pylint: disable=broad-except
+            # We intentionally WANT to catch all exceptions here
+            message = "Unable to generate or encrypt data key using {}".format(trace_info)
+            _LOGGER.exception(message)
+            raise EncryptKeyError(message)
+
+        encryption_materials.add_encrypted_data_key(
+            encrypted_data_key=encrypted_key, keyring_trace=KeyringTrace(wrapping_key=trace_info, flags=_ENCRYPT_FLAGS)
+        )
+
+        return encryption_materials
+
+    def on_decrypt(self, decryption_materials, encrypted_data_keys):
+        # type: (DecryptionMaterials, Iterable[EncryptedDataKey]) -> DecryptionMaterials
+        for edk in encrypted_data_keys:
+            if decryption_materials.data_encryption_key is not None:
+                return decryption_materials
+
+            if not (
+                edk.key_provider.provider_id == _PROVIDER_ID
+                and edk.key_provider.key_info.decode("utf-8") == self._key_id
+            ):
+                continue
+
+            decryption_materials = _try_aws_kms_decrypt(
+                client_supplier=self._client_supplier,
+                decryption_materials=decryption_materials,
+                grant_tokens=self._grant_tokens,
+                encrypted_data_key=edk,
+            )
+
+        return decryption_materials
+
+
+@attr.s
+class _AwsKmsDiscoveryKeyring(Keyring):
+    """AWS KMS discovery keyring that will attempt to decrypt any AWS KMS encrypted data key.
+
+    This keyring should never be used directly.
+    It should only ever be used internally by :class:`KmsKeyring`.
+
+    .. versionadded:: 1.5.0
+
+    :param ClientSupplier client_supplier: Client supplier to use when asking for clients
+    :param List[str] grant_tokens: AWS KMS grant tokens to include in requests (optional)
+    """
+
+    _client_supplier = attr.ib(validator=instance_of(ClientSupplier))
+    _grant_tokens = attr.ib(
+        default=attr.Factory(tuple),
+        validator=(deep_iterable(member_validator=instance_of(six.string_types)), value_is_not_a_string),
+    )
+
+    def on_encrypt(self, encryption_materials):
+        # type: (EncryptionMaterials) -> EncryptionMaterials
+        return encryption_materials
+
+    def on_decrypt(self, decryption_materials, encrypted_data_keys):
+        # type: (DecryptionMaterials, Iterable[EncryptedDataKey]) -> DecryptionMaterials
+        for edk in encrypted_data_keys:
+            if decryption_materials.data_encryption_key is not None:
+                return decryption_materials
+
+            if edk.key_provider.provider_id != _PROVIDER_ID:
```

Comment:
as with the comment above,
maybe what I'm thinking is that a filter would be positive logic,
which would be easier to read?

Reply:
Nice change to positive.